README.md
<a href="http://tarantool.org">
   <img src="https://avatars2.githubusercontent.com/u/2344919?v=2&s=250"
align="right">
</a>

# HTTPS-server for Tarantool
## Table of contents
* [General information](#general-information)
* [Installation](#installation)
* [API](#api)
* [An example of using the module](#an-example-of-using-the-module)

## General information
Link to [GitHub](https://github.com/a1div0/https "GitHub").
Used extern libs:
* [ACME-client](https://github.com/a1div0/acme-client)
* [SSL-socket](https://github.com/tarantool/sslsocket)
* [HTTP-server](https://github.com/tarantool/http)

## Installation
You can:
* clone the repository:
``` shell
git clone https://github.com/a1div0/https.git
```
* install the `https` module using `tarantoolctl`:
```shell
tarantoolctl rocks install https://raw.githubusercontent.com/a1div0/https/main/https-1.0.3-1.rockspec
```

## API
* `local https = require('https.server')` - acquire a library handle
* `local server = https.new(options)` - create new server-object
* `server:route(options, proc)` - binds address and handler
* `server:start()` - starts the server
* `server:stop()` - stop the server
* `server:schedule()` - procedure to run on schedule - reissue the certificate
when its expiration date comes up

### new
```
new(options)
```
This procedure create a new server object with the specified parameters.
The `options` parameter, which is a table with fields:
* `host` - the host to bind to
* `port80` - the port 80 to bind to
* `port443` - the port 443 to bind to
* `dns_name` - site name for certificate
* `cert_path` - path to the folder for storing the certificate and for creating
temporary files in the process of obtaining a certificate
* `cert_name` - file name of the prepared certificate, or, how to name the
certificate when receiving it

For more options see [acme-client](https://github.com/a1div0/acme-client).

### serverObject:route
It is possible to automatically route requests between different handlers,
depending on the request path. For more information see
[http-server](https://github.com/tarantool/http).

### serverObject:start and serverObject:stop
Starts and stop the server.

### serverObject:schedule
Checks if the deadline has passed, and if so, starts the process of obtaining a
new certificate and restarts the server with a new certificate.

## An example of using the module
``` lua
    local https = require('https.server')
    
    local function echo_proc(request)
        local response = request:render{ text = request.body }
        response.headers['x-test-header'] = request.method..' '..request.path;
        response.status = 200
        return response
    end
    
    local options = {
        host = '0.0.0.0',
        port80 = 80,
        port443 = 443,
        dns_name = 'mysite.me',
        cert_path = 'cert/',
        cert_name = 'ssl.pem',
    }
    
    local server = https.new(options)
    server:route({ path = '/echo' }, echo_proc)
    server:start()
```

CHANGELOG
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.3] - 2022-02-13
### Fixed
- Fix bug in `no SSL` mode

## [1.0.2] - 2022-02-01
### Added
- Easy mode (no SSL)

## [1.0.0] - 2022-02-01
The first version has been implemented

test/ide-debug.lua
package.cpath = package.cpath .. ';test/?.so'
local dbg = require('emmy_core')
dbg.tcpListen('localhost', 9999)
dbg.waitIDE()
--dbg.breakHere()

return dbg

test/server.test.lua
require('test.ide-debug')
local tap = require('tap')
local https_lib = require('https.server')
local http_client = require('http.client')
local acme_lib = require("acme-client")
local test = tap.test('https.redirect tests')

local function test_main(test)

    test:plan(12)

    local function echo_proc(request)
        local body = request:read()
        local response = request:render{ text = body }
        response.headers['x-test-header'] = request.method..' '..request.path;
        response.status = 200
        return response
    end

    local options = {
        host = '127.0.0.1',
        port80 = 8080,
        port443 = 443,
        dns_name = 'vpoint.me',
        cert_path = '/srv/sftp/sftp-user/https/test/cert/',
        cert_name = 'ssl-test.pem',
        ssl = false,
    }
    local cert_full_name = options.cert_path..options.cert_name
    local echo_path = '/echo'
    local echo_url = ''
    local server = nil

    os.remove(cert_full_name)

    if options.port80 == 80 then
        echo_url = 'http://'..options.host..echo_path
    else
        echo_url = string.format('http://%s:%d%s', options.host, options.port80, echo_path)
    end

    server = https_lib.new(options)
    server:route({ path = echo_path }, echo_proc)
    server:start()

    local r = http_client.post(echo_url, 'TEST0', {timeout=3})
    test:is(r.status, 200, 'Check easy mode (no SSL) - status 200')
    test:is(r.body, 'TEST0', 'Check easy mode (no SSL) - body echo')

    server:stop()

    if options.port443 == 443 then
        echo_url = 'https://'..options.dns_name..echo_path
    else
        echo_url = string.format('https://%s:%d%s', options.dns_name, options.port443, echo_path)
    end

    local server = https_lib.new(options)
    test:isnt(server, nil, 'HTTPS-server create without cert-file')

    server:route({ path = echo_path }, echo_proc)
    server:start()

    local cert_time1 = acme_lib.certValidTo(cert_full_name)
    test:isnt(cert_time1, nil, 'Check first SSL-certificate')
    require("fiber").sleep(1)
    
    local r = http_client.post(echo_url, 'TEST1', {timeout=3})
    test:is(r.status, 200, 'Status response 1')
    test:is(r.body, 'TEST1', 'Body response 1')

    server.time_to_reissue = 2*(cert_time1 - os.time())
    server:schedule()

    local cert_time2 = acme_lib.certValidTo(cert_full_name)
    test:isnt(cert_time2, nil, 'Check second SSL-certificate')
    test:ok(cert_time1 < cert_time2, 'Check reissue SSL-certificate over active SSL-channel')

    r = http_client.post(echo_url, 'TEST2', {timeout=3})
    test:is(r.status, 200, 'Status response 2')
    test:is(r.body, 'TEST2', 'Body response 2')

    server:stop()

    local r = http_client.post(echo_url, 'TEST3', {timeout=3})
    test:isnt(r.status, 200, 'Status response 3')
    local reason = string.lower(r.reason)
    local timeout_pos = reason:find("timeout", 1, true)
    test:isnt(timeout_pos, nil, 'Reason response 3')

end

test:plan(1)
test:test('https.redirect main test', test_main)

os.exit(test:check() == true and 0 or -1)

test/redirect.test.lua
--require("test.ide-debug")
local lib = require("https.redirect")
local socket = require("socket")
local tap = require('tap')
local test = tap.test('https.redirect tests')

local host = "127.0.0.1"
local port80 = 8080

local client_data = [[POST /api/catalog_items-get-new? HTTP/1.1
Accept: */*
Host: vpoint.me
Sec-fetch-dest: empty
Sec-fetch-site: same-origin
Accept-language: ru-RU,ru;q=0.9,en-US;q=0.8,en;q=0.7

Hello!
]]

local check_server_data = [[HTTP/1.1 301 Found
Location: https://vpoint.me/api/catalog_items-get-new?
]]

test:plan(1)

test:test('https.redirect main test', function(test)
    test:plan(3)

    local r = lib.new()
    r:start(host, port80)

    local test_socket, err = socket.tcp_connect(host, port80)
    test:isnt(test_socket, nil, "Client socket create")
    if test_socket == nil then
        print("ERROR: "..err)
    end

    test_socket:send(client_data)
    test_socket:wait()
    local server_data = test_socket:recv()
    test_socket:close()

    server_data = server_data:gsub("\r\n", "\n")
    test:is(server_data, check_server_data, "Check receive data")

    r:stop()

    test_socket, err = socket.tcp_connect(host, port80)
    test:is(test_socket, nil, "Check stopped module")
end)

os.exit(test:check() == true and 0 or -1)

https/server.lua
local acme_lib = require('acme-client')
local fio = require('fio')
local http_server_lib = require('http.server')
local https_redirect_lib = require('https.redirect')
local sslsocket = require('sslsocket')

local function update_valid_flags(self)
    local diff = self.cert_valid_to - os.time()
    self.cert_need_reissue = self.options.ssl and (diff < self.time_to_reissue)
    self.cert_is_valid = diff >= 3 * 60
end

local function update_valid_to(self)
    local f = nil

    if self.options.ssl then
        f = fio.open(self.cert_full_name, {"O_RDONLY"})
    end

    if f then
        f:close()
        local ok, valid_to_time = pcall(acme_lib.certValidTo, self.cert_full_name)
        if ok then
            self.cert_valid_to = valid_to_time
            update_valid_flags(self)
        end
    else
        self.cert_need_reissue = self.options.ssl
        self.cert_is_valid = false
    end
end

local function route(self, options, proc)
    table.insert(self.route_table, {options = options, proc = proc})
end

local function ssl_listen(host, port, options)
    local cert = options.http_server.cert_full_name
    if cert == nil then
        error('Must be define `cert_full_name` property!')
    end

    local ctx = sslsocket.ctx(sslsocket.methods.tlsv12)
    local rc = sslsocket.ctx_use_private_key_file(ctx, cert)
    if rc == false then
        error('Private key is invalid: '..cert)
    end

    rc = sslsocket.ctx_use_certificate_file(ctx, cert)
    if rc == false then
        error('Certificate is invalid: '..cert)
    end

    return sslsocket.tcp_server(host, port, options.handler, nil, ctx)
end

local function start_ssl(self)
    if (self.ssl_enable) then
        self.redirect = https_redirect_lib.new()
        self.redirect:start(self.options.host, self.options.port80)

        self.server = http_server_lib.new(self.options.host, self.options.port443)
        self.server.tcp_server_f = ssl_listen
        self.server.cert_full_name = self.cert_full_name

        self.ssl_active = true
    else
        self.server = http_server_lib.new(self.options.host, self.options.port80)
    end

    for _, tuple in ipairs(self.route_table) do
        self.server:route(tuple.options, tuple.proc)
    end

    self.server:start()
end

local function stop_ssl(self)
    self.ssl_active = false
    self.redirect:stop()
    self.server:stop()
end

local function setup_challenge_proc(self)
    local server_handle = self.server
    local function setup_challenge_http01(url, body)
        local proc = nil
        if body ~= nil then
            proc = function (request)
                return request:render{status = 200, text = body}
            end
        else
            proc = function (request)
                return request:render{status = 404}
            end
        end
        server_handle:route({ path = url }, proc)
    end
    return setup_challenge_http01
end

local function start(self)

    if self.ssl_enable then
        update_valid_to(self)

        if self.cert_need_reissue then
            self.server = http_server_lib.new(self.options.host, self.options.port80)
            self.server:start()
            local proc = setup_challenge_proc(self)
            local acme_client = acme_lib.new(self.options, proc)
            acme_client:getCert()
            self.server:stop()
            update_valid_to(self)
        end
    end

    start_ssl(self)
end

local function stop(self)
    if self.ssl_active then
        stop_ssl(self)
    else
        self.server:stop()
    end
end

local function schedule(self)
    update_valid_flags(self)
    if self.cert_need_reissue then
        self:stop()
        self:start()
    end
end

local exports = {
    new = function(options)
        local self = {
            time_to_reissue = 24 * 3600,
            options = options,
            route_table = {},
            route = route,
            start = start,
            stop = stop,
            schedule = schedule,
            ssl_enable = options.ssl ~= false
        }

        self.options.internalIP4 = options.host or options.internalIP4
        self.options.dnsName = options.dns_name or options.dnsName
        self.options.certPath = options.cert_path or options.certPath
        self.options.certName = options.cert_name or options.certName
        self.cert_full_name = self.options.certPath..self.options.certName

        return self
    end
}

return exports

https/redirect.lua
local socket = require('socket')

local function get_new_url(sock)
    local line1 = sock:read("\n")
    local method, url, version = line1:match("(.+) (.+) (.+)\n")

    local findPhrase = "Host: "
    local loop = true
    while loop do
        local line = sock:read("\n")
        if line == nil then
            break
        else
            local pos = line:find(findPhrase, 1, true)
            if pos == 1 then
                local host = line:sub(pos + string.len(findPhrase), -2)
                return "https://"..host..url
            end
        end
    end
end

local function handler(sock)
    local ok, result = pcall(get_new_url, sock)
    if ok and result ~= nil and type(result) == "string" then
        sock:write("HTTP/1.1 301 Found\r\n")
        sock:write("Location: "..result.."\r\n")
    else
        sock:write("HTTP/1.1 500 Internal Server Error\r\n")
        sock:write("Content-Length: 10\r\n")
        sock:write("Content-Type: text/plain\r\n")
        sock:write("\r\n")
        if (result ~= nil) then
            sock:write(""..result)
        else
            sock:write("Bad request")
        end
    end
end

local function start(self, host, port80)
    self.s = socket.tcp_server(host, port80, handler)
end

local function stop(self)
    if self.s ~= nil then
        self.s:close()
    end
end

local exports = {
    new = function()
        local self = {
            start = start,
            stop = stop,
        }

        return self
    end
}

return exports